### PR TITLE
Enable the allocations during concurrent sweep - Part2.

### PR DESCRIPTION
**Pull request by atulkatti**
If allocations are to be allowed to existing heap blocks during concurrent sweep, we set aside a few
heap blocks from the heapBlockList prior to beginning sweep. However, we eed to then go back and make sure these blocks also swept before this sweep finishes. In order to do this we clearly define concurrent sweep having 2 passes now. These passes existed before but were not distiguished as they would always start and finish in one go on the background thread. However, whenever allocations are allowed during concurrent sweep; the concurrent sweep will start Pass1 on the background thread, wait to finish Pass1 of the blocks we set aside to allocate from on the main thread and then go back to finish Pass2 for all heap blocks on the background thread. Note that, due to this need to finish Pass1 on the foreground thread the overall background sweep will now appear to take longer whenever we chose to do such a two-pass sweep.

The sequence of things we do to allow allocations during concurrent sweep is described below:
1. At the beginning of concurrrent sweep we decide if we will benefit from allowing allocations during concurrent sweep for any of the buckets. If there is at-least one bucket for which we think we will benefit we will turn on allocations during concurrent sweep. Once turned on we will attempt to enable allocations during concurrent sweep for all supported buckets (i.e. small/medium, normal/leaf, non-finalizable buckets.write barrrier bickets are supported as well.).
2. If allocations are turned on during concurrent sweep, we will see if there are any allocable blocks in the
heapBlockList after the nextAllocableBlockHead. If we find any such blocks, we move them to a SLIST that the allocator can pick these blocks from during sweep.
3. CollectionStateConcurrentSweepPass1: We will finish Pass1 of the sweep for all the remaining blocks (other than the ones we put in the SLIST in step 2 above) This will generally happen on the background thread unless we are forcing in-thread sweep. This state is now specifically identified as CollectionStateConcurrentSweepPass1.
4. CollectionStateConcurrentSweepPass1Wait: At this point we need to wait for all the blocks that we put in the SLIST to also finish the Pass1 of the sweep. This needs to happen on the foreground thread so we prevent the allocator from picking up the blocks from SLIST while we do this. This state is now identified as CollectionStateConcurrentSweepPass1Wait.
5. CollectionStateConcurrentSweepPass2: At this point we will do the actual sweeping of all the blocks that are not yet swept, for eaxample, any blocks that were put onto the pendingSweepList. As these blocks get swept we keep adding them to the SLIST again to allow allocators to allocate from them as soon as they are swept.
6. Before the Pass2 can finish we can call this concurrent sweep done we need to move all the blocks off of the SLIST so that normal allocations can begin after the sweep. This is the last step of the concurrent sweep.

### CODE REVIEW

File: lib/Common/Memory/HeapBlock.inl [hunk 1]
```diff
@@ -17,6 +17,17 @@ SmallHeapBlockT<TBlockAttributes>::SetAttributes(void * address, unsigned char a
...
+template <class TBlockAttributes>
+void
+SmallHeapBlockT<TBlockAttributes>::UpdateAttributes(void * address, unsigned char attributes)
+{
+    Assert(this->address != nullptr);
+    Assert(this->segment != nullptr);
+    ushort index = GetAddressIndex(address);
+    Assert(index != SmallHeapBlockT<TBlockAttributes>::InvalidAddressBit);
+    ObjectInfo(index) = attributes;
+}
```
Comment (atulkatti) [inline lib/Common/Memory/HeapBlock.inl]: This is unused. I will get rid of this.

File: lib/Common/CommonDefines.h [hunk 2]
```diff
@@ -168,6 +168,7 @@
 #define SYSINFO_IMAGE_BASE_AVAILABLE 1
 #define ENABLE_CONCURRENT_GC 1
 #define ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP 1 // Only takes effect when ENABLE_CONCURRENT_GC is enabled.
+#define ALLOCATIONS_DURING_CONCURRENT_SWEEP_HEAPBLOCK_THRESHOLD 5000 // Number of blocks a heap bucket needs to have before allocations during concurrent sweep feature kicks-in.
```
Comment (digitalinfinity) [inline lib/Common/CommonDefines.h]: Why is this defined here and not in RecyclerHeuristic.h?

Since you're touching this code, can you refactor the defines slightly? You can probably move ENABLE_CONCURRENT_GC outside of the _WIN32 define, and just check for #if ENABLE_VALGRIND?

Also, why is allocation during concurrent sweep disabled on xplat? This feature would benefit node scenarios...

File: lib/Common/Memory/CollectionState.h [hunk 3]
```diff
@@ -39,6 +39,12 @@ enum CollectionState
     Collection_PostSweepRedeferralCallback = 0x00040000,
     Collection_WrapperCallback             = 0x00080000,
 
+#if ENABLE_CONCURRENT_GC && ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+    Collection_ConcurrentSweepPass1 = 0x00100000,
```
Comment (digitalinfinity) [inline lib/Common/Memory/CollectionState.h]: Can we have more descriptive names other than Pass1 and Pass2? What actually happens to a heap block during Pass1 sweep and what happens during Pass2 sweep? At the very least, we should have that documented here if we can't come up with better names

File: lib/Common/Memory/HeapBlock.h [hunk 4]
```diff
@@ -105,15 +105,15 @@ enum ObjectInfoBits : unsigned short
     EnumClass_1_Bit             = 0x01,    // This can be extended to add more enumerable classes (if we still have bits left)
 
     // Mask for above bits
-    StoredObjectInfoBitMask     = 0xFF,
+    StoredObjectInfoBitMask = 0xFF,
```
Comment (digitalinfinity) [inline lib/Common/Memory/HeapBlock.h]: Nit- preserve the existing alignment of assignments for everything in this enum

File: lib/Common/Memory/HeapBlock.h [hunk 5]
```diff
@@ -941,6 +963,32 @@ class HeapBlockList
         return tail;
     }
 
+    template <typename TBlockType>
+    static TBlockType * FindPreviousBlock(TBlockType * list, TBlockType * heapBlockToFind)
```
Comment (digitalinfinity) [inline lib/Common/Memory/HeapBlock.h]: Is this still used?

File: lib/Common/Memory/HeapBlock.cpp [hunk 6]
```diff
@@ -187,7 +187,7 @@ SmallHeapBlockT<TBlockAttributes>::ConstructorCommon(HeapBucket * bucket, ushort
     this->Init(objectSize, objectCount);
     Assert(heapBlockType < HeapBlock::HeapBlockType::SmallAllocBlockTypeCount + HeapBlock::HeapBlockType::MediumAllocBlockTypeCount);
     Assert(objectCount > 1 && objectCount == (this->GetPageCount() * AutoSystemInfo::PageSize) / objectSize);
-#ifdef RECYCLER_SLOW_CHECK_ENABLED
+#if defined(RECYCLER_SLOW_CHECK_ENABLED) || (ENABLE_CONCURRENT_GC && ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP)
```
Comment (digitalinfinity) [inline lib/Common/Memory/HeapBlock.cpp]: Doesn't ENABLE_ALLOCATIONS_DURING_CONCCURRENT_SWEEEP => ENABLE_CONCURRENT_GC? When would the allocations be enabled but not concurrent gc?

File: lib/Common/Memory/Recycler.inl [hunk 7]
```diff
@@ -543,7 +543,10 @@ Recycler::NotifyFree(T * heapBlock)
         this->isForceSweeping = true;
         heapBlock->isForceSweeping = true;
 #endif
-        heapBlock->template SweepObjects<SweepMode_InThread>(this);
+#ifdef RECYCLER_TRACE
+        this->PrintBlockStatus(nullptr, heapBlock, _u("[**34**] calling SweepObjects during NotifyFree."));
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.inl]: What does 34 here mean?

File: lib/Common/Memory/Recycler.cpp [hunk 8]
```diff
@@ -160,6 +164,9 @@ Recycler::Recycler(AllocationPolicyManager * policyManager, IdleDecommitPageAllo
     enableConcurrentMark(false),  // Default to non-concurrent
     enableParallelMark(false),
     enableConcurrentSweep(false),
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+    allocationsDuringConcurrentSweepTurnedOn(false),
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: As I understand this, this is whether it's turned on for a particular collection, not for the thread context lifetime, correct? Would be good to rename this to make that clear, otherwise i dont understand whats the difference between this and the config flag

File: lib/Common/Memory/SmallHeapBlockAllocator.h [hunk 9]
```diff
@@ -173,6 +178,41 @@ SmallHeapBlockAllocator<TBlockType>::InlinedAllocImpl(Recycler * recycler, DECLS
...
         }
 #endif
+#if ENABLE_CONCURRENT_GC && ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+        if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc) && recycler->AllocationsDuringconcurrentSweepTurnedOn())
+        {
+            // If we are allocating during concurrent sweep we must mark the object to prevent it from being swept
+            // in the ongoing sweep.
+            if (heapBlock != nullptr && heapBlock->isPendingConcurrentSweepPrep)
+            {
+                AssertMsg(!this->isAllocatingFromNewBlock, "We shouldn't be tracking allocation to a new block; i.e. bump allcoation; during concurrent sweep.");
```
Comment (digitalinfinity) [inline lib/Common/Memory/SmallHeapBlockAllocator.h]: Nit: typo

File: lib/Common/Memory/SmallHeapBlockAllocator.h [hunk 10]
```diff
@@ -173,6 +178,41 @@ SmallHeapBlockAllocator<TBlockType>::InlinedAllocImpl(Recycler * recycler, DECLS
...
+                AssertMsg(!heapBlock->IsAnyFinalizableBlock(), "Allocations are not allowed to finalizable blocks during concurrent sweep.");
+                AssertMsg(heapBlock->heapBucket->AllocationsStartedDuringConcurrentSweep(), "We shouldn't be allocating from this block while allocations are disabled.");
+
+                // Explcitly mark this object and also clear the free bit.
+                uint bitIndex = heapBlock->GetAddressBitIndex(memBlock);
+                Assert(heapBlock->IsValidBitIndex(bitIndex));
+
+                heapBlock->GetMarkedBitVector()->Set(bitIndex);
```
Comment (digitalinfinity) [inline lib/Common/Memory/SmallHeapBlockAllocator.h]: This seems like an excessive amount of cost for these allocations- couple of ideas:
- Snapshot the markbit vector before adding it to the SLIST, and then set just the mark bit here. When you need to compute the bits for the new objects, you can just do mark - markSnapshot.. The count of that is the number of objects allocated during concurrent sweep so you don't need to keep track of that separately either. You can then unset those bits from the free bit vector/debug free bit vector to fix up both those bit vectors.

**Comment by digitalinfinity** [comment]
>     bool IsFreeListAllocMode() const

Does this imply allocating from a blog during concurrent sweep? For this to be true, the heap block would have to be not null, which is true in your case.

Also, it might be worth adding another function which is IsAllocatingDuringConcurrentSweepMode

---
Refers to: lib/Common/Memory/SmallHeapBlockAllocator.h:65 in eea291e. [](commit_id = eea291e743461729a2d2fd07f6ec72f1b43f9893, deletion_comment = False)

File: lib/Common/Memory/SmallHeapBlockAllocator.h [hunk 11]
```diff
@@ -173,6 +178,41 @@ SmallHeapBlockAllocator<TBlockType>::InlinedAllocImpl(Recycler * recycler, DECLS
             Assert(isSet);
         }
 #endif
+#if ENABLE_CONCURRENT_GC && ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+        if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc) && recycler->AllocationsDuringconcurrentSweepTurnedOn())
```
Comment (digitalinfinity) [inline lib/Common/Memory/SmallHeapBlockAllocator.h]: Lots of conditions and i wonder if they're redundant?

File: lib/Common/Memory/SmallHeapBlockAllocator.h [hunk 12]
```diff
@@ -173,6 +178,41 @@ SmallHeapBlockAllocator<TBlockType>::InlinedAllocImpl(Recycler * recycler, DECLS
...
+
+                DebugOnly(heapBlock->objectsMarkedDuringSweep++);
+                // We need to keep track of the number of objects allocated during concurrent sweep, to be
+                // able to make the correct determination about whether a block is EMPTY or FULL when the actual
+                // sweep of this block happens.
+                heapBlock->objectsAllocatedDuringConcurrentSweepCount++;
+#ifdef RECYCLER_TRACE
+                if (recycler->GetRecyclerFlagsTable().Trace.IsEnabled(Js::ConcurrentSweepPhase) && recycler->GetRecyclerFlagsTable().Trace.IsEnabled(Js::MemoryAllocationPhase))
+                {
+                    Output::Print(_u("[**33**]FreeListAlloc: Object 0x%p from HeapBlock 0x%p used for allocation during ConcurrentSweep [CollectionState: %d] \n"), memBlock, heapBlock, recycler->collectionState);
```
Comment (digitalinfinity) [inline lib/Common/Memory/SmallHeapBlockAllocator.h]: Nit: what's 33?

File: lib/Common/Memory/Recycler.h [hunk 13]
```diff
@@ -784,8 +787,9 @@ class Recycler
     DListBase<GuestArenaAllocator> guestArenaList;
     DListBase<ArenaData*> externalGuestArenaList;    // guest arenas are scanned for roots
 
-#ifdef RECYCLER_PAGE_HEAP
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.h]: Why was this moved?

File: lib/Common/Memory/Recycler.h [hunk 14]
```diff
@@ -1653,6 +1658,14 @@ class Recycler
     {
         return ((collectionState & Collection_ConcurrentSweep) == Collection_ConcurrentSweep);
     }
+
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+    bool AllocationsDuringconcurrentSweepTurnedOn()
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.h]: Nit: camel case name here- and I think AllocAllocationDuringConcurrentSweep is a tad bit more compact

File: lib/Common/Memory/Recycler.cpp [hunk 15]
```diff
@@ -3072,15 +3079,36 @@ Recycler::Sweep(bool concurrent)
...
 
-           DoBackgroundWork(true);
-           // Continue as if the concurrent sweep were executing
-           // Next time we check for completion, we will finish the sweep just as if it had happened out of thread.
+        if(needForceForground)
+        {
+            // Failed to spawn the concurrent sweep.
+            // Instead, force the concurrent sweep to happen right here in thread.
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+            if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc) && this->AllocationsDuringconcurrentSweepTurnedOn())
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: What does this mean? We're sweeping here in the foreground because we failed to spawn the background sweep so why do the two-pass sweep here?

File: lib/Common/Memory/Recycler.cpp [hunk 16]
```diff
@@ -4358,7 +4390,11 @@ BOOL
 Recycler::RequestConcurrentWrapperCallback()
 {
 #if ENABLE_CONCURRENT_GC
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+    Assert(!IsConcurrentExecutingState() && !IsConcurrentSweepState());
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Why this change? It looks like it's sufficient to just check for IsConcurrentExecutingState since that implies we're not in concurrent sweep state, no?

File: lib/Common/Memory/Recycler.cpp [hunk 17]
```diff
@@ -4464,12 +4500,16 @@ Recycler::FinishConcurrent()
 
         const BOOL forceFinish = flags & CollectOverride_ForceFinish;
 
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+        if (forceFinish || !(IsConcurrentExecutingState()))
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Isn't this the same condition as #else?

File: lib/Common/Memory/Recycler.cpp [hunk 18]
```diff
@@ -4522,7 +4562,11 @@ Recycler::TryFinishConcurrentCollect()
     Assert(!concurrent || !forceInThread);
     if (concurrent && concurrentThread != NULL)
     {
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+        if (IsConcurrentExecutingState())
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Ditto- same condition in #if and #else?

File: lib/Common/Memory/Recycler.cpp [hunk 19]
```diff
@@ -4629,7 +4673,19 @@ Recycler::IsConcurrentSweepSetupState() const
 BOOL
 Recycler::IsConcurrentSweepState() const
 {
-    return this->collectionState == CollectionStateConcurrentSweep;
+#if ENABLE_CONCURRENT_GC && ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+    if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc))
+    {
+        return this->collectionState == CollectionStateConcurrentSweepPass1 ||
+            this->collectionState == CollectionStateConcurrentSweepPass1Wait ||
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Interesting- before, we'd be in ConcurrentSweepState if there was work going on in the background thread- now, this could be true even if we're in the foreground thread?

File: lib/Common/Memory/Recycler.cpp (hunk 16 above)
Comment (digitalinfinity) [reply lib/Common/Memory/Recycler.cpp]: Actually, I guess its a little more subtle- the question is, can this function get called if we're midway during concurrent sweep? And it sounds like the answer is no? Although in theory, that should not be a problem since this function can only be called during Mark/Rescan, so I guess broadening the assert is probably fine- and there's no reason to ifdef it here, IsConcurrentSweepState should do the right thing in both cases

---
In reply to: [148170498](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148170498) [](ancestors = 148170498)

File: lib/Common/Memory/Recycler.cpp [hunk 20]
```diff
@@ -5674,8 +5731,57 @@ Recycler::FinishConcurrentCollect(CollectionFlags flags)
...
+            FinishTransferSwept(flags);
+        }
+        else
+        {
+            // Signal the background thread to finish concurrent sweep Pass2 for all the buckets.
+            SetEvent(this->concurrentWorkReadyEvent);
+            needConcurrentSweep = true;
+        }
+    }
+    //else if (collectionState == CollectionStateConcurrentSweepPass2Wait)
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Remove commented code?

File: lib/Common/Memory/Recycler.cpp [hunk 21]
```diff
@@ -5674,8 +5731,57 @@ Recycler::FinishConcurrentCollect(CollectionFlags flags)
...
+#ifdef RECYCLER_TRACE
+            if (this->GetRecyclerFlagsTable().Trace.IsEnabled(Js::ConcurrentSweepPhase))
+            {
+                Output::Print(_u("[GC #%d] Finishing Sweep Pass2 in-thread. \n"), this->collectionCount);
+            }
+#endif
+            this->recyclerSweep->FinishSweep();
+            this->FinishConcurrentSweep();
+            this->recyclerSweep->EndBackground();
+            //this->collectionState = CollectionStateConcurrentSweepPass2Wait;
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Remove comment code

File: lib/Common/Memory/Recycler.cpp [hunk 22]
```diff
@@ -5674,8 +5731,57 @@ Recycler::FinishConcurrentCollect(CollectionFlags flags)
...
+            this->collectionState = CollectionStateTransferSweptWait;
+            RECYCLER_PROFILE_EXEC_BACKGROUND_END(this, Js::ConcurrentSweepPhase);
+
+            FinishTransferSwept(flags);
+        }
+        else
+        {
+            // Signal the background thread to finish concurrent sweep Pass2 for all the buckets.
+            SetEvent(this->concurrentWorkReadyEvent);
+            needConcurrentSweep = true;
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Probably should set this before SetEvent?

File: lib/Common/Memory/Recycler.cpp [hunk 23]
```diff
@@ -5674,8 +5731,57 @@ Recycler::FinishConcurrentCollect(CollectionFlags flags)
 #endif
         GCETW_INTERNAL(GC_STOP, (this, ETWEvent_ConcurrentRescan));
     }
+#if ENABLE_CONCURRENT_GC && ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+    else if (collectionState == CollectionStateConcurrentSweepPass1Wait)
+    {
+        this->FinishSweepPrep();
+        this->collectionState = CollectionStateConcurrentSweepPass2;
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: I'd move this into the if. And I'd change the else condition to call StartConcurrent instead of SetEvent

File: lib/Common/Memory/Recycler.cpp [hunk 24]
```diff
@@ -5873,54 +5979,127 @@ Recycler::DoBackgroundWork(bool forceForeground)
...
+                {
+                    Output::Print(_u("[GC #%d] Finishing Sweep Pass2 on background thread. \n"), this->collectionCount);
+                }
+#endif
+#if ENABLE_BACKGROUND_PAGE_ZEROING
+                if (CONFIG_FLAG(EnableBGFreeZero))
+                {
+                    // Drain the zero queue again as we might have free more during sweep
+                    // in the background
+                    GCETW(GC_BACKGROUNDZEROPAGE_START, (this));
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: Why are you doing this before we finish the sweep? You're going to do this anyway below, no?

File: lib/Common/CommonDefines.h (hunk 2 above)
Comment (leirocks) [reply lib/Common/CommonDefines.h]: I guess that's because we don't have SLIST support on linux...

---
In reply to: [148163317](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148163317) [](ancestors = 148163317)

File: lib/Common/Memory/HeapBucket.h [hunk 25]
```diff
@@ -75,8 +75,12 @@ class HeapBucket
 protected:
     HeapInfo * heapInfo;
     uint sizeCat;
+    bool allocationsStartedDuringConcurrentSweep;
```
Comment (leirocks) [inline lib/Common/Memory/HeapBucket.h]: >allocationsStartedDuringConcurrentSweep [](start = 9, length = 39)

why this is not in the following #if?

File: lib/Common/Memory/CollectionState.h [hunk 26]
```diff
@@ -63,6 +69,12 @@ enum CollectionState
 
     CollectionStateSetupConcurrentSweep   = Collection_Sweep | Collection_ConcurrentSweepSetup,               // setting up concurrent sweep
     CollectionStateConcurrentSweep        = Collection_ConcurrentSweep | Collection_ExecutingConcurrent,      // concurrent sweep
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+    CollectionStateConcurrentSweepPass1 = Collection_ConcurrentSweep | Collection_ConcurrentSweepPass1 | Collection_ExecutingConcurrent,          // concurrent sweep Pass 1
+    CollectionStateConcurrentSweepPass1Wait = Collection_ConcurrentSweep | Collection_ConcurrentSweepPass1Wait,                                   // concurrent sweep wait state after Pass 1 has finished
+    CollectionStateConcurrentSweepPass2 = Collection_ConcurrentSweep | Collection_ConcurrentSweepPass2 | Collection_ExecutingConcurrent,          // concurrent sweep Pass 2
+    CollectionStateConcurrentSweepPass2Wait = Collection_ConcurrentSweep | Collection_ConcurrentSweepPass2Wait | Collection_ExecutingConcurrent,  // concurrent sweep wait state after Pass 2 has finished
```
Comment (curtisman) [inline lib/Common/Memory/CollectionState.h]: >Collection_ExecutingConcurrent [](start = 113, length = 30)

Collection_ExecutingConcurrent shouldn't be here?

File: lib/Common/Memory/SmallHeapBlockAllocator.h (hunk 12 above)
Comment (atulkatti) [reply lib/Common/Memory/SmallHeapBlockAllocator.h]: see my response above. 

---
In reply to: [148169799](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148169799) [](ancestors = 148169799)

File: lib/Common/CommonDefines.h (hunk 2 above)
Comment (atulkatti) [reply lib/Common/CommonDefines.h]: Good point, moved the heuristic to RecyclerHeuristic.h. Lei is right about the SLIST support. We will look at adding this support in future. I will try to do the refactoring if I can now.

---
In reply to: [148163317](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148163317) [](ancestors = 148163317)

File: lib/Common/Memory/CollectionState.h (hunk 3 above)
Comment (atulkatti) [reply lib/Common/Memory/CollectionState.h]: I will keep the names and add some documentation here as well. I have added details about the 2-pass sweep in the documentation for the method PrepareForAllocationsDuringConcurrentSweep.

---
In reply to: [148163410](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148163410) [](ancestors = 148163410)

File: lib/Common/Memory/HeapBlock.h (hunk 5 above)
Comment (atulkatti) [reply lib/Common/Memory/HeapBlock.h]: nope. removed.

---
In reply to: [148163808](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148163808) [](ancestors = 148163808)

File: lib/Common/Memory/HeapBlock.cpp (hunk 6 above)
Comment (atulkatti) [reply lib/Common/Memory/HeapBlock.cpp]: You are right. I could only check for ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP. I will simplify this everywhere.

---
In reply to: [148167429](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148167429) [](ancestors = 148167429)

File: lib/Common/Memory/Recycler.inl (hunk 7 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.inl]: :) It doesn't have any particular significance, apart from the fact that the lower numbers mean early stages of the sweep and higher numbers mean later stages. It only tells me where in code this step happened. Across multiple GCs as heap blocks go through a journey I found this useful in finding out if things are making progress the way we expect. This kind of tracing will be enabled only in verbose mode to diagnose/understand how certain type of heap blocks are processed during sweep.

---
In reply to: [148167948](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148167948) [](ancestors = 148167948)

File: lib/Common/Memory/SmallHeapBlockAllocator.h (hunk 11 above)
Comment (atulkatti) [reply lib/Common/Memory/SmallHeapBlockAllocator.h]: I have removed the config flag check as it is now redundant.

---
In reply to: [148169494](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148169494) [](ancestors = 148169494)

File: lib/Common/Memory/Recycler.cpp (hunk 24 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.cpp]: Even for one pass sweep we try to do this before and after the sweep to make any queued pages available for allocation. I wanted to be consistent with that. Let me know if you think otherwise.

---
In reply to: [148171629](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148171629) [](ancestors = 148171629)

**Comment by atulkatti** [comment]
>     bool IsFreeListAllocMode() const

No, IsFreeListAllocMode() means this is not a "new" block i.e. we are not still in bump alloc mode for this block. This doesn't necessarily mean we are allocating during concurrent sweep. I will add the new function, good suggestion.

---
In reply to: [340957983](https://github.com/Microsoft/ChakraCore/pull/4104#issuecomment-340957983) [](ancestors = 340957983)

---
Refers to: lib/Common/Memory/SmallHeapBlockAllocator.h:65 in eea291e. [](commit_id = eea291e743461729a2d2fd07f6ec72f1b43f9893, deletion_comment = False)

File: lib/Common/Memory/Recycler.h (hunk 13 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.h]: I believe I hit a build error at some point in some config. I will undo this change. It seems alright now.

---
In reply to: [148169931](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148169931) [](ancestors = 148169931)

File: lib/Common/Memory/Recycler.cpp (hunk 15 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.cpp]: Good catch. We will just need to make sure that the feature is turned off in this case. I will make this change.

---
In reply to: [148170163](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148170163) [](ancestors = 148170163)

File: lib/Common/Memory/Recycler.cpp (hunk 16 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.cpp]: good point. fixed.

---
In reply to: [148171246](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148171246) [](ancestors = 148171246,148170498)

File: lib/Common/Memory/Recycler.cpp (hunk 17 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.cpp]: good catch. fixed.

---
In reply to: [148170532](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148170532) [](ancestors = 148170532)

File: lib/Common/Memory/Recycler.cpp (hunk 19 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.cpp]: Yes, that's right. I think it should be OK because the two pass sweep does do processing in foreground thread as well.

---
In reply to: [148170831](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148170831) [](ancestors = 148170831)

File: lib/Common/Memory/Recycler.cpp (hunk 22 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.cpp]: Shouldn't matter. I have moved it up anyways.


---
In reply to: [148171378](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148171378) [](ancestors = 148171378)

File: lib/Common/Memory/HeapBucket.h (hunk 25 above)
Comment (atulkatti) [reply lib/Common/Memory/HeapBucket.h]: It can be. I will do that change.

---
In reply to: [148181880](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148181880) [](ancestors = 148181880)

File: lib/Common/Memory/CollectionState.h (hunk 26 above)
Comment (atulkatti) [reply lib/Common/Memory/CollectionState.h]: good catch. fixed. 

---
In reply to: [148390821](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148390821) [](ancestors = 148390821)

File: lib/Common/CommonDefines.h (hunk 2 above)
Comment (obastemur) [reply lib/Common/CommonDefines.h]: Just FYI; locks may eventually involve for xplat SLIST. There are platform specific weirdness with interlocked exchange utilities that we may not depend on our favorite arch.

Last time I remember the potential gain from having the SLIST wasn’t big hence I had reduced the priority significantly. Is this something we may reconsider SLIST xplat priority?

File: lib/Common/Memory/HeapBlock.cpp [hunk 27]
```diff
@@ -187,7 +187,7 @@ SmallHeapBlockT<TBlockAttributes>::ConstructorCommon(HeapBucket * bucket, ushort
...
     heapBucket->heapInfo->heapBlockCount[heapBlockType]++;
```
Comment (curtisman) [inline lib/Common/Memory/HeapBlock.cpp]: Where is this used in release bits for your feature?

File: lib/Common/Memory/SmallNormalHeapBucket.cpp [hunk 28]
```diff
@@ -401,17 +448,59 @@ SmallNormalHeapBucketBase<TBlockType>::SweepPartialReusePages(RecyclerSweep& rec
...
+                    Assert(!this->IsAnyFinalizableBucket());
+                    bool blockAddedToSList = HeapBucketT<TBlockType>::PushHeapBlockToSList(this->allocableHeapBlockListHead, heapBlock);
+
+                    // If we encountered OOM while pushing the heapBlock to the SLIST we must add it to the heapBlockList so we don't lose track of it.
+                    if (!blockAddedToSList)
+                    {
+#ifdef RECYCLER_TRACE
+                        this->GetRecycler()->PrintBlockStatus(this, heapBlock, _u("[**10**] finished SweepPartialReusePages, heapblock REUSED but OOM while adding to the SLIST."));
+#endif
+                        this->GetRecycler()->OutOfMemory();
```
Comment (curtisman) [inline lib/Common/Memory/SmallNormalHeapBucket.cpp]: You can't logically continue at this point.  OOM may not always FailFast. 

We should find a way to recover from this case just like before.

File: lib/Common/Memory/HeapBlock.cpp (hunk 27 above)
Comment (atulkatti) [reply lib/Common/Memory/HeapBlock.cpp]: Good catch. We don't use this but the count on the heap bucket. I will undo this change.

---
In reply to: [148563397](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148563397) [](ancestors = 148563397)

File: lib/Common/Memory/SmallNormalHeapBucket.cpp (hunk 28 above)
Comment (atulkatti) [reply lib/Common/Memory/SmallNormalHeapBucket.cpp]: Ok. I will change this to fail fast for now and will plan to work on the correct thing to do here later.

---
In reply to: [148566503](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148566503) [](ancestors = 148566503)

File: lib/Common/Memory/Recycler.cpp (hunk 8 above)
Comment (atulkatti) [reply lib/Common/Memory/Recycler.cpp]: Yes, this applies only for the current GC. I will rename the field to indicate that.

---
In reply to: [148168140](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148168140) [](ancestors = 148168140)

File: lib/Common/CommonDefines.h (hunk 2 above)
Comment (atulkatti) [reply lib/Common/CommonDefines.h]: Yes, Oz, it would be great if we can get the SLIST work done sooner. Let's sync up on that offline.

---
In reply to: [148497285](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148497285) [](ancestors = 148497285)

**Comment by matthargett** [comment]
Is this PR waiting on something in particular before it gets merged?

**Comment by atulkatti** [comment]
@matthargett  I am reviewing some test results. This is expected to be merged soon.
